File: backend/app/tests/factories/participant_factory.py
import random
import uuid

from faker import Faker

from app.schemas import ParticipantCreate

faker = Faker()

STUDY_GROUPS = ["treatment", "control"]
STATUSES = ["active", "completed", "withdrawn"]
GENDERS = ["M", "F", "Other"]

_subject_counter = 0


def _next_subject_id() -> str:
    global _subject_counter
    _subject_counter += 1
    return f"P{_subject_counter:03d}"


def generate_create_participant_dto(**overrides) -> ParticipantCreate:
    return ParticipantCreate(
        subject_id=overrides.get("subject_id", _next_subject_id()),
        study_group=overrides.get("study_group", random.choice(STUDY_GROUPS)),
        enrollment_date=overrides.get(
            "enrollment_date", faker.date_between(start_date="-1y", end_date="today").isoformat()
        ),
        status=overrides.get("status", random.choice(STATUSES)),
        age=overrides.get("age", faker.random_int(min=18, max=80)),
        gender=overrides.get("gender", random.choice(GENDERS)),
    )


def generate_participant_dict(**overrides) -> dict:
    dto = generate_create_participant_dto(**overrides)
    return {
        "participant_id": str(uuid.uuid4()),
        **dto.model_dump(),
        "enrollment_date": dto.enrollment_date.isoformat(),
    }

File: frontend/src/test/participant.factory.ts
import { faker } from '@faker-js/faker';

import type { CreateParticipantDto, Participant } from '@/schemas/participantSchema';

export interface IGenerateParticipant {
  status?: Participant['status'];
  study_group?: Participant['study_group'];
  gender?: Participant['gender'];
}

export function generateParticipant(overrides?: IGenerateParticipant): Participant {
  return {
    participant_id: faker.string.uuid(),
    subject_id: `P${faker.number.int({ min: 100, max: 999 })}`,
    study_group: overrides?.study_group ?? faker.helpers.arrayElement(['treatment', 'control']),
    enrollment_date: faker.date.recent({ days: 365 }).toISOString().split('T')[0],
    status: overrides?.status ?? faker.helpers.arrayElement(['active', 'completed', 'withdrawn']),
    age: faker.number.int({ min: 18, max: 80 }),
    gender: overrides?.gender ?? faker.helpers.arrayElement(['M', 'F', 'Other']),
  };
}

export function generateCreateParticipantDto(
  overrides?: IGenerateParticipant,
): CreateParticipantDto {
  const { participant_id: _, ...rest } = generateParticipant(overrides);
  return rest;
}
